crates/router/src/core/payments.rs

```diff
@@ -507,7 +520,160 @@ where
+    connector_details: Option<&api::ConnectorCallType>,
+) -> Option<String> {
+    connector_details.and_then(|connector_call_type| match connector_call_type {
+        api::ConnectorCallType::Single(data) => Some(data.connector_name.to_string()),
+        _ => None,
+    })
+}
+
+fn is_payment_method_tokenization_enabled_for_connector(
+    state: &AppState,
+    connector_name: &str,
+    payment_method: &storage::enums::PaymentMethod,
+) -> RouterResult<bool> {
+    let connector_tokenization_filter = state.conf.tokenization.0.get(connector_name);
+
+    Ok(connector_tokenization_filter
+        .map(|connector_filter| connector_filter.payment_method.contains(payment_method))
+        .unwrap_or(false))
+}
+
+async fn decide_payment_method_tokenize_action(
+    state: &AppState,
+    connector_name: &str,
+    payment_method: &storage::enums::PaymentMethod,
+    pm_parent_token: Option<&String>,
+    is_connector_tokenization_enabled: bool,
+) -> TokenizationAction {
+    match pm_parent_token {
+        None => {
+            if is_connector_tokenization_enabled {
+                TokenizationAction::TokenizeInConnectorAndRouter
+            } else {
+                TokenizationAction::TokenizeInRouter
+            }
+        }
+        Some(token) => {
+            let redis_conn = connection::redis_connection(&state.conf).await;
+            let key = format!(
+                "pm_token_{}_{}_{}",
+                token.to_owned(),
+                payment_method,
+                connector_name
+            );
+            // We can't discard the error, we need to log it
+            logger::debug!(
+                connector_token_redis_error=?redis_conn.get_key::<String>(&key).await.err()
+            );
+
+            match redis_conn.get_key::<String>(&key).await.ok() {
+                None => {
+                    if is_connector_tokenization_enabled {
+                        TokenizationAction::TokenizeInConnector
+                    } else {
+                        TokenizationAction::SkipConnectorTokenization
+                    }
+                }
+                Some(connector_token) => TokenizationAction::ConnectorToken(connector_token),
+            }
+        }
+    }
+}
+
+pub enum TokenizationAction {
+    TokenizeInRouter,
+    TokenizeInConnector,
+    TokenizeInConnectorAndRouter,
+    ConnectorToken(String),
+    SkipConnectorTokenization,
+}
+
+#[allow(clippy::too_many_arguments)]
+pub async fn get_connector_tokenization_action<F, Req>(
+    state: &AppState,
+    operation: &BoxedOperation<'_, F, Req>,
+    connector_details: Option<&api::ConnectorCallType>,
+    mut payment_data: PaymentData<F>,
+    validate_result: &operations::ValidateResult<'_>,
+    _req: &Req,
+) -> RouterResult<(PaymentData<F>, TokenizationAction)>
+where
+    F: Send + Clone,
+{
+    let payment_data_and_tokenization_action =
+        if get_connector_from_connector_type(connector_details).is_some()
```

Comment:
You can use 
```rust
match get_connector_from_connector_type(connector_details) {
	Some(connector) if is_operation_confirm(&operation)
}
```
Instead of calling `get_connector_from_connector_type` again in line number 615